src/main/java/algorithms/curated170/medium/maze2/TheMazeIIDijkstra.java

```diff
@@ -0,0 +1,77 @@
+package algorithms.curated170.medium.maze2;
+
+import java.util.Arrays;
+import java.util.PriorityQueue;
+
+ 
+public class TheMazeIIDijkstra {
+ 
+    int[][] maze;
+  public int shortestDistance(int[][] maze, int[] start, int[] dest) {
+ 
+        this.maze = maze;
+        
+        return dijkstra(start, dest); 
+    }
+   
+ public int dijkstra( int[] start, int[] destination) {
+         
+        PriorityQueue<int[]> pq = new PriorityQueue<>((p1, p2) -> p1[2] - p2[2]); 
+        int[][] directions = {{0, -1}, {1, 0}, {-1, 0}, {0, 1}};
+     
+        pq.add(new int[]{start[0], start[1], 0});
+        while (!pq.isEmpty()) {
+        
+            int[] startPoint = pq.poll();
+           
+            if (startPoint[0] < 0)
+                break;
+            
+            int x = startPoint[0];
+            int y = startPoint[1];
+            
+            if(maze[x][y]==-1) continue;
+            
+            maze[x][y] = -1;   
+            if (x == destination[0] && y == destination[1]) {
+                return startPoint[2];
+            }
+           
+       
+           for (int[] dir: directions) {
+                int dirx = dir[0];
+                int diry = dir[1];
+           
+                int nextX = x;
+                int nextY = y;
+               
+                int currentDistance =  startPoint[2];
+                
+                while (canPass(nextX, nextY)) {
+                    nextX += dirx;
+                    nextY += diry;
+                    currentDistance++;
+                }
+
+                nextX -= dirx;
+                nextY -= diry;
+                currentDistance--;
+           
+                
+                 if(maze[nextX][nextY]!=-1) {
+                      
+                      pq.add(new int[]{nextX, nextY, currentDistance});
+                }
+            }
+        }
+        return -1;
+    }
+ 
+     boolean canPass(int nextX, int nextY) {
+            return (nextX >= 0 && nextY >= 0
+                    && nextX < maze.length
+                    && nextY < maze[0].length
+                    && maze[nextX][nextY] != 1);
```

Comment:
Why don't we check for this condition unlike in the previous two solutions?
`maze[nextX][nextY] == 0);`

Reply:
I have understood the reason why we don't do this.
In the previous solution, the tiles that we come across are either empty or a wall. If empty, we move through them. If wall, we stop. Here, we have a three valued logic: A tile is either empty but visited, empty or a wall. We should be able to move through an empty tile even if it is visited. At the end, we can't stop and branch off there, but whilst moving to another valid tile, we should be able to pass them.

Reply:
Thanks for the explanation Erdem. Appreciated.